resources/type-system/flow-analysis-v0.md

```diff
@@ -0,0 +1,339 @@
+  - If `VM1 = VariableModel(d1, p1, s1, a1, u1, c1)` and
+  - If `VM2 = VariableModel(d2, p2, s2, a2, u2, c2)` then
+  - `VM3 = VariableModel(d3, p3, s3, a3, u3, c3)` where
+   - `d3 = d1 = d2`
+     - Note that all models must agree on the declared type of a variable
+   - `p3 = p1 ^ p2`
+     - `p1` and `p2` are totally ordered subsets of a global partial order.
+  Their intersection is a subset of each, and as such is also totally ordered.
+   - `s3 = s1 U s2`
+     - The set of test sites is the union of the test sites on either path
+   - `a3 = a1 && a2`
+     - A variable is definitely assigned in the join of two models iff it is
+       definitely assigned in both.
+   - `u3 = u1 && u2`
+     - A variable is definitely unassigned in the join of two models iff it is
+       definitely unassigned in both.
+   - `c3 = c1 || c2`
+     - A variable is captured in the join of two models iff it is captured in
+       either.
+
+- `split(M)`, where `M = FlowModel(r, VM)` is a flow model which models program
+  nodes inside of a control flow split, and is defined as `FlowModel(r2, VM)`
+  where `r2` is `r` with `true` pushed as the top element of the stack.
+
+- `drop(M)`, where `M = FlowModel(r, VM)` is a flow model which models program
+  nodes after a control flow split which are only reachable by one path through
+  the split, and is defined as `FlowModel(r1, VM)` where `r0 = pop(r)` and `r1 =
+  push(pop(r0), top(r0) && top(r1))`.  Equivalently, `drop(M)` may be thought of
+  as `join(M, M)`.
+
+- `join(M1, M2)`, where `M1` and `M2` are flow models, represents the union of
+  two flow models and is defined as follows:
+
+  We define `join(M1, M2)` where `M1 = FlowModel(r1, VI1)` and `M2 =
+  FlowModel(r2, VI2))` and `pop(r1) = pop(r2) = r0` for some `r0` to be `M3` where:
+    - if `top(r1)` is true and `top(r2)` is false, then `M3` is `FlowModel(r0, VI1)`.
+    - if `r1` is false and `r2` is true, then `M3` is `FlowModel(pop(r2), VI2)`.
+    - otherwise `M3 = FlowModel(r3, VI3)` where:
+      - `r3` is `push(pop(r0), top(r0) && top(r1) && top(r2))`
+      - `VI3` maps each variable `v` in the domain of `VI1` and `VI2` to
+      `joinV(VI1(v), VI2(v))`.  Note that any variable which is in domain of
+      only one of the two is dropped, since it is no longer in scope.
+
+  Both join and joinV are commutative and associative by construction.
+
+  For brevity, we will sometimes extend `join` to more than two arguments in the
+  obvious way.  For example, `join(M1, M2, M3)` represents `join(join(M1, M2),
+  M3)`, and `join(S)`, where S is a set of models, denotes the result of folding
+  all models in S together using `join`.
+
+- `exit(M)` represents the model corresponding to a program location which is
+  unreachable, but is otherwise modeled by flow model `M = FlowModel(r, VI)`,
+  and is defined as `FlowModel(push(pop(r), false), VI)`
+
+### Promotion
+
+Promotion policy is defined by the following operations on flow models.
+
+Policy:
```

Comment:
I've tried to capture the policy options for assignment promotion that were proposed [here](https://github.com/dart-lang/language/issues/440) here.  What other options should I include?